compiler/noirc_evaluator/src/acir/tests/instructions.rs

```diff
@@ -1 +1,302 @@
+}
+
+#[test]
+fn call_with_predicate() {
+    let src = "
+    acir(inline) fn main f0 {
+      b0(v0: Field, v1: u1):
+        enable_side_effects v1
+        v2 = call f1(v0) -> Field
+        return v2
+    }
+
+    acir(fold) fn one f1 {
+      b0(v0: Field):
+        v1 = add v0, Field 1
+        return v1
+    }
+    ";
+    let program = ssa_to_acir_program(src);
+
+    assert_circuit_snapshot!(program, @r"
+    func 0
+    private parameters: [w0, w1]
+    public parameters: []
+    return values: [w2]
+    BLACKBOX::RANGE input: w1, bits: 1
+    CALL func: 1, predicate: w1, inputs: [w0], outputs: [w3]
+    ASSERT w3 = w2
+
+    func 1
+    private parameters: [w0]
+    public parameters: []
+    return values: [w1]
+    ASSERT w1 = w0 + 1
+    ");
+}
+
+#[test]
+fn not() {
+    let src = "
+    acir(inline) fn main f0 {
+      b0(v0: u8):
+        v1 = not v0
+        return v1
+    }
+    ";
+    let program = ssa_to_acir_program(src);
+    assert_circuit_snapshot!(program, @r"
+    func 0
+    private parameters: [w0]
+    public parameters: []
+    return values: [w1]
+    BLACKBOX::RANGE input: w0, bits: 8
+    ASSERT w1 = -w0 + 255
+    ");
+}
+
+#[test]
+fn truncate_u16_to_6_bits() {
+    let src = "
+    acir(inline) fn main f0 {
+      b0(v0: u16):
+        v1 = truncate v0 to 6 bits, max_bit_size: 8
+        return v1
+    }
+    ";
+    let program = ssa_to_acir_program(src);
+
+    // Truncating to 6 bits is the same as dividing by 2^6 = 64
+    assert_circuit_snapshot!(program, @r"
+    func 0
+    private parameters: [w0]
+    public parameters: []
+    return values: [w1]
+    BLACKBOX::RANGE input: w0, bits: 16
+    BRILLIG CALL func: 0, inputs: [w0, 64], outputs: [w2, w3]
+    BLACKBOX::RANGE input: w2, bits: 2
+    BLACKBOX::RANGE input: w3, bits: 6
+    ASSERT w3 = w0 - 64*w2
+    ASSERT w3 = w1
+
+    unconstrained func 0: directive_integer_quotient
+    0: @10 = const u32 2
+    1: @11 = const u32 0
+    2: @0 = calldata copy [@11; @10]
+    3: @2 = field int_div @0, @1
+    4: @1 = field mul @2, @1
+    5: @1 = field sub @0, @1
+    6: @0 = @2
+    7: stop &[@11; @10]
+    ");
+}
+
+#[test]
+fn truncate_field_to_6_bits() {
+    let src = "
+    acir(inline) fn main f0 {
+      b0(v0: Field):
+        v1 = truncate v0 to 6 bits, max_bit_size: 254
+        return v1
+    }
+    ";
+    let program = ssa_to_acir_program(src);
+
+    // Truncating to 8 bits is the same as dividing by 2^6 = 64.
+    // TODO: explain what overflow we try to avoid in the rest of the opcodes
+    assert_circuit_snapshot!(program, @r"
+    func 0
+    private parameters: [w0]
+    public parameters: []
+    return values: [w1]
+    BRILLIG CALL func: 0, inputs: [w0, 64], outputs: [w2, w3]
+    BLACKBOX::RANGE input: w2, bits: 248
+    BLACKBOX::RANGE input: w3, bits: 6
+    ASSERT w3 = w0 - 64*w2
+    ASSERT w4 = -w2 + 342003794872488675347600089769644923258568193756500536620284440415247007744
+    BLACKBOX::RANGE input: w4, bits: 248
```

Comment:
This is our `max_q_bits`. For `r` we have 2^6 = 64 which takes 7 bits to represent. As we are working with a 254 bit field we are in fact `>= F::max_num_bits() - 1`. As per the comment in `euclidean_division_var` we want to assert `q*b+r < 2^max_q_bits*2^max_rhs_bits`.

Reply:
`F::max_num_bits()` is the max num bits of our `F: AcirField`. The check occurs at the end of `euclidean_division_var`.

Reply:
Aaah... got it. So 254 >= 254 - 1 👍

Reply:
nit: We actually have 255 >= 254 - 1